### ISSUE TITLE
No built-in way to access parent command from subcommand?

### ISSUE THREAD

**Issue by michaelpj** [remkop/picocli#247]
A natural thing to do with parent commands is define options that will apply to all the subcommands. However, it's not terribly convenient to access the parent command from the subcommand: you have to rely on whatever creates the `CommandLine` to thread everything around.

Similarly, the provided `IParseResultHandler`s don't make it easy to pass things around.

I think what would be most convenient for me would be for a subcommand to receive an instance of its parent command as a constructor argument. That way everything is well-typed and it's easy to access the options of parent commands.

**Comment by remkop** [comment]
You raise a good point. 

I’m thinking to provide an additional `@ParentCommand` annotation to address this. 

A reference to the parent command will be injected to any fields annotated with `@ParentCommand`.  This also provides type safety while not imposing any constraints on subcommand constructors. 

Would that work for you?

**Comment by michaelpj** [comment]
Yes, that should do the job. 

(Incidentally, I don't think the requirements on constructors (there must be a zero-arg constructor) for commands are documented anywhere.)

**Comment by remkop** [comment]
Ok. Thanks for pointing out the docs can be improved. 

By the way, do zero-arg constructors work for you or are you also interested in using a factory like requested in #169 ?

**Comment by michaelpj** [comment]
Zero-arg constructors are fine for me at the moment, although it would be kind of neat to be able to use a factory.

**Comment by remkop** [comment]
Pushed the new `@ParentCommand` feature to master.
Please verify.

**Comment by remkop** [comment]
Reopening: still need to update the user manual.

**Comment by remkop** [comment]
Note to self:
```
Index: src/main/java/picocli/CommandLine.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/main/java/picocli/CommandLine.java	(revision )
+++ src/main/java/picocli/CommandLine.java	(revision )
@@ -1502,7 +1502,7 @@
      * class Top implements Runnable {
      *
      *     &#064;Option(names = {"-d", "--directory"}, description = "this option applies to all subcommands")
-     *     private File baseDirectory;
+     *     File baseDirectory;
      *
      *     public void run() { System.out.println("Hello from top"); }
      * }
@@ -1513,13 +1513,8 @@
      *     &#064;ParentCommand
      *     private Top parent;
      *
-     *     &#064;Parameters(description = "The number of times to print the result")
-     *     private int count;
-     *
      *     public void run() {
-     *         for (int i = 0; i &lt; count; i++) {
-     *             System.out.println("Subcommand: parent command 'directory' is " + parent.baseDirectory);
-     *         }
+     *         System.out.println("Subcommand: parent command 'directory' is " + parent.baseDirectory);
      *     }
      * }
      * </pre>
```

**Comment by remkop** [comment]
This issue has been addressed in picocli 2.2, which is now available on JCenter and Maven Central. 
See the [release notes](https://github.com/remkop/picocli/releases/tag/v2.2.0) for details.
